### PR TITLE
Let invoice.Encode receive a function which hashes itself the message

### PR DESCRIPTION
**Pull request by yaslama**
Modify the SignCompact function passed to invoice.Encode to receive the
message before it's hashed and hash it itself.
With this modification, the SignMessage rpc function from the signrpc
subserver can be used and an invoice can be encoded from outside lnd.

For instance, in the gist https://gist.github.com/yaslama/e71e50bdf489246a4e0720942393f4a8 a payment request can be generated and signed from another one with all the information being the same as the original payment request except the amount. This can be used to generate a payment request with a different amount than the amount stored in the payee's database. One important use case is 'on the fly' channel creation via a node that provides channel creation services using HtlcInterceptor:
1. The payee generates an invoice with a specific amount and sends it to the payer.
2. An intermediate node providing channel creation services intercepts the request and asks the payee for a new payment request with a smaller amount.
3. The intermediate node settles the original amount and keeps the difference which is the cost of creating a new channel with the payee.

### CODE REVIEW

**Comment by bhandras** [comment]
Could you explain how this is different than just hashing the arbitrary message and then signing the hash?

**Comment by yaslama** [comment]
> Could you explain how this is different from just hashing the arbitrary message and then signing the hash?

I don't know if I understand your question, but the final signature is the same. What is different is that with this PR, it's possible to call invoice.Encode(signer) where signer is a function using  signrpc from lnd. Without the PR, I didn't find a way to sign an invoice from a lnd client (connecting to lnd using grpc).

**Comment by bhandras** [comment]
> > Could you explain how this is different from just hashing the arbitrary message and then signing the hash?
> 
> I don't know if I understand your question, but the final signature is the same. What is different is that with this PR, it's possible to call invoice.Encode(signer) where signer is a function using signrpc from lnd. Without the PR, I didn't find a way to sign an invoice from a lnd client (connecting to lnd using grpc).

Ok, I now understand what you meant.

**Comment by yaslama** [comment]
> needs rebase, otherwise LGTM 100

Thanks! I just rebased the PR.

File: zpay32/encode.go [hunk 1]
```diff
@@ -95,7 +94,7 @@ func (invoice *Invoice) Encode(signer MessageSigner) (string, error) {
 				"signature: %v", err)
 		}
 
-		valid := signature.Verify(hash, invoice.Destination)
+		valid := signature.Verify(chainhash.HashB(toSign), invoice.Destination)
```
Comment (halseth) [inline zpay32/encode.go:97]: line looks long

File: zpay32/invoice.go [hunk 2]
```diff
@@ -100,11 +100,11 @@ var (
 // MessageSigner is passed to the Encode method to provide a signature
 // corresponding to the node's pubkey.
 type MessageSigner struct {
-	// SignCompact signs the passed hash with the node's privkey. The
-	// returned signature should be 65 bytes, where the last 64 are the
+	// SignCompact hashes and signs the passed msg with the node's privkey.
```
Comment (halseth) [inline zpay32/invoice.go:103]: maybe better: `SignCompact signs the hash of the passed msg...`?

File: lnrpc/invoicesrpc/addinvoice.go [hunk 3]
```diff
@@ -387,7 +388,9 @@ func AddInvoice(ctx context.Context, cfg *AddInvoiceConfig,
 
 	payReqString, err := payReq.Encode(
 		zpay32.MessageSigner{
-			SignCompact: cfg.NodeSigner.SignDigestCompact,
+			SignCompact: func(msg []byte) ([]byte, error) {
+				return cfg.NodeSigner.SignDigestCompact(chainhash.HashB(msg))
```
Comment (halseth) [inline lnrpc/invoicesrpc/addinvoice.go:392]: long line

File: zpay32/invoice_test.go [hunk 4]
```diff
@@ -915,8 +915,8 @@ func TestInvoiceChecksumMalleability(t *testing.T) {
 
 	privKey, _ := btcec.PrivKeyFromBytes(btcec.S256(), privKeyBytes)
 	msgSigner := MessageSigner{
-		SignCompact: func(hash []byte) ([]byte, error) {
-			return btcec.SignCompact(btcec.S256(), privKey, hash, true)
+		SignCompact: func(msg []byte) ([]byte, error) {
+			return btcec.SignCompact(btcec.S256(), privKey, chainhash.HashB(msg), true)
```
Comment (halseth) [inline zpay32/invoice_test.go:919]: long line

**Comment by yaslama** [comment]
@halseth Thanks for your review.

**Comment by Roasbeef** [comment]
Needs a rebase so this can land in time for 0.12!

**Comment by yaslama** [comment]
> Needs a rebase so this can land in time for 0.12!
I rebased it. The PR is ready for the landing!!

**Comment by halseth** [comment]
Somehow not getting picked up by travis... @yaslama could you try rebasing it again? :P

**Comment by yaslama** [comment]
> Somehow not getting picked up by travis... @yaslama could you try rebasing it again? :P

@halseth It seems that the failure in travis is not related to the PR itself not the rebasing. Can you trigger a new travis check?

**Pull request merged**